### PR TITLE
Create an LP contract 

### PR DESCRIPTION
**Pull request by LOUDINISouad**

### CODE REVIEW

**Comment by LOUDINISouad** [comment]
This code faces a problem with the gas fees and some bugs when I try to test it , I still can not fix it !! @

File: src/LP.sol [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        totalLiquidity += liquidityMinted;
+    }
+
+    function withdraw(uint256 liquidityAmount) external {
+        require(liquidity[msg.sender] >= liquidityAmount, "Insufficient liquidity balance");
+
+        uint256 amountA = liquidityAmount *(tokenA.balanceOf(address(this)))/(totalLiquidity);
+        uint256 amountB = liquidityAmount *(tokenB.balanceOf(address(this)))/(totalLiquidity);
+
+        liquidity[msg.sender] = liquidity[msg.sender]-(liquidityAmount);
```
Comment (achab) [inline src/LP.sol:39]: `liquidity[msg.sender] -= liquidityAmount;`

File: src/LP.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        totalLiquidity = totalLiquidity -(liquidityAmount);
```
Comment (achab) [inline src/LP.sol:40]: `totalLiquidity -= liquidityAmount;`

File: src/LP.sol [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+contract LiquidityPool {
+    using Math for uint256;
+
+    IERC20 public tokenA;
+    IERC20 public tokenB;
+    address public owner;
+    uint256 public totalLiquidity;
+    mapping(address => uint256) public liquidity;
+    uint256 public fee = 30; // 0.30% fee
```
Comment (achab) [inline src/LP.sol:15]: we'd rather use capital letters for constants

`uint256 public constant FEE = 30;`

File: src/LP.sol [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
```
Comment (achab) [inline src/LP.sol:16]: you could add another constant to represent 1 in bips i.e. 10000 (and not 1000 like you put in your code).

`uint256 public constant ONE_IN_BPS = 1e4;`

File: src/LP.sol [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    function swap(uint256 amountAIn) external {
+        require(amountAIn > 0, "Amount must be greater than 0");
+
+        uint256 amountBOut = calculateSwapAmount(amountAIn);
+        require(tokenA.transferFrom(msg.sender, address(this), amountAIn), "Failed to transfer tokenA for swap");
+        require(tokenB.transfer(msg.sender, amountBOut), "Failed to transfer tokenB for swap");
+    }
+
+    function calculateLiquidityAmount(uint256 amountA, uint256 amountB) public pure returns (uint256) {
+        uint256 adjustedAmountA = amountA*(997); // 997 corresponds to 1000 - fee (0.30%)
```
Comment (achab) [inline src/LP.sol:55]: `uint256 adjustedAmountA = amountA * (ONE_IN_BPS - FEE);`

File: src/LP.sol [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        uint256 adjustedAmountB = amountB*(997); // 997 corresponds to 1000 - fee (0.30%)
```
Comment (achab) [inline src/LP.sol:56]: ``uint256 adjustedAmountB = amountB * (ONE_IN_BPS - FEE);``

File: src/LP.sol [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        return  Math.sqrt(adjustedAmountA*(adjustedAmountB));
+    }
+
+    function calculateSwapAmount(uint256 amountAIn) internal view returns (uint256) {
+        uint256 amountAWithFee = amountAIn*(1000); // 1000 corresponds to 1000 (100%)
```
Comment (achab) [inline src/LP.sol:61]: where are the fees here?

File: src/LP.sol [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        uint256 numerator = amountAWithFee*(tokenB.balanceOf(address(this)));
+        uint256 denominator = tokenA.balanceOf(address(this))*(1000)+(amountAWithFee);
```
Comment (achab) [inline src/LP.sol:63]: replace `1000` with `ONE_IN_BPS`

**Review by achab — APPROVED** [review]

**Pull request merged**